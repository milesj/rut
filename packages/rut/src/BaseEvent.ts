export default class BaseEvent<T = unknown> {
  bubbles: boolean = true;

  cancelable: boolean = true;

  currentTarget!: T;

  defaultPrevented: boolean = false;

  eventPhase: number = 0;

  isTrusted: boolean = true;

  propagationStopped: boolean = false;

  srcElement: null = null;

  target!: T;

  timeStamp: number;

  type: string;

  constructor(type: string) {
    this.timeStamp = Date.now();
    this.type = type;
  }

  preventDefault(): void {
    this.defaultPrevented = true;
  }

  stopImmediatePropagation(): void {
    this.propagationStopped = true;
  }

  stopPropagation(): void {
    this.propagationStopped = true;
  }
}
