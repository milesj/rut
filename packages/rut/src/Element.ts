/* eslint-disable @typescript-eslint/no-unused-vars */

import React from 'react';
import { ReactTestInstance } from 'react-test-renderer';
import { getTypeName } from './internals/react';
import { doAct, doAsyncAct } from './internals/act';
import { debug } from './internals/debug';
import { getPropForDispatching } from './internals/element';
import { whereTypeAndProps } from './predicates';
import {
  Predicate,
  DispatchOptions,
  DebugOptions,
  UnknownProps,
  AtIndexType,
  QueryOptions,
  AdapterRendererOptions,
  ElementType,
} from './types';

export default abstract class Element<
  Type extends ElementType = ElementType,
  Props = never,
  Host = unknown
> {
  // @ts-expect-error Set after instantiation
  options: AdapterRendererOptions;

  protected element: ReactTestInstance;

  private readonly isRutElement = true;

  constructor(element: ReactTestInstance) {
    this.element = element;
  }

  /**
   * Log and return a JSX representation of the current *reconciled* React component tree.
   * Does not include exotic components or nodes, such as:
   *
   *  - Context consumers and providers
   *  - Roots, portals, modes
   *  - Profiler, Suspense
   *  - Fragments
   */
  debug = (options: DebugOptions = {}) => debug(this.element, options);

  /**
   * Dispatch an event listener for the defined prop name. Requires a synthetic event
   * based on the original event type.
   *
   * Note: This may only be executed on host components.
   */
  dispatch(name: string, eventOrConfig?: unknown, options: DispatchOptions = {}): this {
    const prop = getPropForDispatching(this, name);
    const event = this.createSyntheticEvent(name, eventOrConfig, this.element.type);

    // istanbul ignore next
    if (options.propagate) {
      // eslint-disable-next-line no-console
      console.warn('Event propagation is experimental and is not fully implemented.');
    }

    doAct(() => prop(event), this.options.applyPatches);

    return this;
  }

  /**
   * Like `dispatch` but also awaits the event so that async calls have time to finish.
   */
  async dispatchAndWait(
    name: string,
    eventOrConfig?: unknown,
    options: DispatchOptions = {},
  ): Promise<void> {
    const prop = getPropForDispatching(this, name);
    const event = this.createSyntheticEvent(name, eventOrConfig, this.element.type);

    // istanbul ignore next
    if (options.propagate) {
      // eslint-disable-next-line no-console
      console.warn('Event propagation is experimental and is not fully implemented.');
    }

    await doAsyncAct(() => prop(event), this.options.applyPatches);
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If any are found, a list of `Element`s is returned.
   */
  find(type: ElementType, props?: UnknownProps): Element<ElementType>[] {
    return this.query(whereTypeAndProps(type, props));
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If any are found, return the `Element` at the defined
   * index. Accepts shorthand `first` and `last` indices.
   */
  findAt(type: ElementType, at: AtIndexType, props?: UnknownProps): Element<ElementType> {
    const results = this.query(whereTypeAndProps(type, props));
    let index: number;

    if (at === 'first') {
      index = 0;
    } else if (at === 'last') {
      index = results.length - 1;
    } else if (typeof at === 'number') {
      index = at;
    } else {
      throw new TypeError(`Invalid index type "${at}".`);
    }

    const value = results[index];

    if (!value) {
      throw new Error(
        `Expected to find an element at index ${index} for \`${getTypeName(type)}\`.`,
      );
    }

    return value;
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If exactly 1 is found, a `Element`s is returned,
   * otherwise an error is thrown.
   */
  findOne(type: ElementType, props?: UnknownProps): Element<ElementType> {
    const results = this.find(type, props);

    if (results.length !== 1) {
      throw new Error(
        `Expected to find 1 element for \`${getTypeName(type)}\`, found ${results.length}.`,
      );
    }

    return results[0];
  }

  /**
   * Return the name of the component, element, or node as JSX.
   */
  name(jsx: boolean = false): string {
    // Use the raw fiber types for names, as they preserve the internal node structures
    const name = getTypeName(this.element._fiber.elementType || this.element._fiber.type);

    return jsx ? `<${name} />` : name;
  }

  /**
   * Search through the current child tree using a custom predicate, which is passed the
   * ReactTestRenderer node and internal React fiber node. If any are found,
   * a list of `Element`s is returned.
   */
  query<T extends ElementType>(predicate: Predicate, options?: QueryOptions): Element<T>[] {
    return this.element
      .findAll((node) => predicate(node, node._fiber), { deep: true, ...options })
      .map((node) => {
        const element = this.options.createElement(node);

        element.options = this.options;

        return element;
      });
  }

  /**
   * Return a ref associated with the current component.
   * If a ref is stored on the internal fiber node, it will be returned.
   * Otherwise an instance, callback, or string ref may be referenced by name.
   */
  ref<T>(name?: string): T | null {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const inst = this.element.instance;
    const fiber = this.element._fiber;

    if (!name || fiber.ref) {
      return fiber.ref;
    }

    // Callback refs
    if (name in inst) {
      return inst[name];
    }

    // String refs
    if (inst.refs[name]) {
      return inst.refs[name];
    }

    return null;
  }

  /**
   * Return name when cast as a string.
   */
  toString(): string {
    return this.name(true);
  }

  /**
   * Create a synthetic event specific to this renderer.
   */
  abstract createSyntheticEvent(
    name: string,
    eventOrConfig: unknown,
    type: ElementType,
  ): React.SyntheticEvent;
}
