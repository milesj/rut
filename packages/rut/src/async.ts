/* eslint-disable @typescript-eslint/no-explicit-any */

const NativePromise = (global.Promise as any) as PromiseConstructor;
const methodsToWrap: (keyof PromiseConstructor)[] = ['all', 'race', 'resolve', 'reject'];

export function wrapAndCaptureAsync(): Promise<any>[] {
  const queue: Promise<any>[] = [];

  function MockPromise<T>(
    executor: (
      resolve: (value?: T | PromiseLike<T>) => void,
      reject: (reason?: any) => void,
    ) => void,
  ): Promise<T> {
    const promise = new NativePromise(executor);

    queue.push(promise);

    return promise;
  }

  methodsToWrap.forEach(method => {
    Object.defineProperty(MockPromise, method, {
      value: (...args: unknown[]) => {
        // @ts-ignore
        const promise = NativePromise[method](...args);

        queue.push(promise);

        return promise;
      },
    });
  });

  global.Promise = MockPromise;

  return queue;
}

export function unwrapPromise() {
  global.Promise = NativePromise;
}

export function waitForAsyncQueue(queue: Promise<any>[]) {
  unwrapPromise();

  return NativePromise.all(queue);
}
