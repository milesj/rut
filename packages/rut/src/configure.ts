import { RendererOptions, IntegrationOptions } from './types';

export const globalOptions: RendererOptions = {};

export function configure(options: RendererOptions) {
  Object.assign(globalOptions, options);
}

export const integrationOptions: IntegrationOptions = {
  runWithTimers(cb) {
    return cb();
  },
};

export function configureIntegration(options: IntegrationOptions) {
  Object.assign(integrationOptions, options);
}
