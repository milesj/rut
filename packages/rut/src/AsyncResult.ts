import Result from './Result';
import { RendererOptions } from './types';
import { doAsyncAct } from './internals/act';
import Element from './Element';

export default class AsyncResult<
  Props extends object = {},
  Root extends Element = Element
> extends Result<Props, Root> {
  /**
   * Like `rerender` but also awaits the re-render so that async calls have time to finish.
   */
  rerenderAndWait = async (element: React.ReactElement<Props>, options?: RendererOptions) => {
    Object.assign(this.options, options);

    await doAsyncAct(
      () =>
        this.renderer.unstable_flushSync(() => this.renderer.update(this.updateElement(element))),
      this.options.applyPatches,
    );

    return this.root;
  };
}
