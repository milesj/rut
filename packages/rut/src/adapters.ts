import React from 'react';
import AsyncResult from './AsyncResult';
import BaseEvent from './BaseEvent';
import Element from './Element';
import { doAct, doAsyncAct } from './internals/act';
import { globalOptions } from './internals/config';
import SyncResult from './SyncResult';
import SyntheticEvent from './SyntheticEvent';
import { AdapterRendererOptions } from './types';

export { BaseEvent, Element, SyntheticEvent };

export function doRender<Props extends object, Root extends Element>(
  element: React.ReactElement<Props>,
  options: AdapterRendererOptions,
): SyncResult<Props, Root> {
  return doAct(
    () =>
      new SyncResult(element, {
        ...globalOptions,
        ...options,
      }),
    options.applyPatches,
  );
}

export async function doRenderAndWait<Props extends object, Root extends Element>(
  element: React.ReactElement<Props>,
  options: AdapterRendererOptions,
): Promise<AsyncResult<Props, Root>> {
  return doAsyncAct(
    () =>
      new AsyncResult(element, {
        ...globalOptions,
        ...options,
      }),
    options.applyPatches,
  );
}
