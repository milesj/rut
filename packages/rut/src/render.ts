import React from 'react';
import AsyncResult from './AsyncResult';
import SyncResult from './SyncResult';
import { globalOptions } from './configure';
import { doAct, doAsyncAct } from './internals/act';
import { RendererOptions } from './types';

// export function render<Props extends object = {}>(
//   element: React.ReactElement<Props>,
//   options?: RendererOptions,
// ): SyncResult<Props> {
//   const opts = {
//     ...globalOptions,
//     ...options,
//   };

//   return doAct(() => new SyncResult(element, opts));
// }

export async function renderAndWait<Props extends object = {}>(
  element: React.ReactElement<Props>,
  options?: RendererOptions,
): Promise<AsyncResult<Props>> {
  const opts = {
    ...globalOptions,
    ...options,
  };

  return doAsyncAct(() => new AsyncResult(element, opts), opts.asyncMode);
}

export const render = renderAndWait;
