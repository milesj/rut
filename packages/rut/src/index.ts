/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { act } from 'react-test-renderer';
import matchers from './matchers';
import { render, renderAndWait } from './render';
import mockSyntheticEvent from './mocks/event';

export * from './helpers';
export * from './types';

export { act, matchers, mockSyntheticEvent, render, renderAndWait };
