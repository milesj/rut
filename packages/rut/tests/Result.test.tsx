import React, { useEffect, useLayoutEffect, useContext } from 'react';
import Element from '../src/Element';
import { render, renderAndWait } from '../src/render';
import {
  ClassComp,
  FuncComp,
  TestContext,
  AsyncHookComp,
  AsyncCdmComp,
  AsyncCduComp,
  TimerCdmComp,
  TimerHookComp,
  TimerCduComp,
  TestProps,
  AsyncProps,
} from './fixtures';

describe('Result', () => {
  const oldWarn = console.warn;
  let warnSpy: jest.Mock;

  beforeEach(() => {
    warnSpy = jest.fn();
    console.warn = warnSpy;
  });

  afterEach(() => {
    console.warn = oldWarn;
  });

  function Wrapper({ children }: { children?: React.ReactNode }) {
    return <TestContext.Provider value="wrapped">{children || null}</TestContext.Provider>;
  }

  it('can serialize as a snapshot', () => {
    const result = render<TestProps>(
      <FuncComp>
        <b>Child</b>
      </FuncComp>,
    );

    expect(result).toMatchSnapshot();
  });

  it('wraps with `StrictMode` when using `strict`', () => {
    class StrictComp extends React.Component {
      componentWillMount() {
        // Logs a warning to be UNSAFE
      }

      render() {
        return null;
      }
    }

    const result = render<{}>(<StrictComp />, { strict: true });

    expect(result).toMatchSnapshot();
    expect(warnSpy).toHaveBeenCalledWith(
      expect.stringContaining('componentWillMount has been renamed'),
    );
  });

  it('wraps with another element when using `wrapper`', () => {
    function ContextComp() {
      const value = useContext(TestContext);

      return <div>{value}</div>;
    }

    const result = render<{}>(<ContextComp />, { wrapper: <Wrapper /> });

    expect(result).toMatchSnapshot();
    expect(result.debug({ log: false })).toContain('<Wrapper>');
    expect(result.root).toContainNode('wrapped');
  });

  it('wraps with both `strict` and `wrapper`', () => {
    class StrictComp extends React.Component {
      componentWillReceiveProps() {
        // Logs a warning to be UNSAFE
      }

      render() {
        return null;
      }
    }

    function Wrapped() {
      return (
        <div>
          <StrictComp />
        </div>
      );
    }

    const result = render<{}>(<Wrapped />, { strict: true, wrapper: <Wrapper /> });

    expect(result).toMatchSnapshot();
    expect(result.debug({ log: false })).toContain('<Wrapper>');
    expect(warnSpy).toHaveBeenCalledWith(
      expect.stringContaining('componentWillReceiveProps has been renamed'),
    );
  });

  it('catches and rethrows errors', () => {
    // React logs when an error is thrown without a boundary
    const spy = jest.spyOn(console, 'error');

    function ErrorComp({ fail }: { fail: boolean }) {
      if (fail) {
        throw new Error('Failed for some reason...');
      }

      return null;
    }

    expect(() => {
      render<{}>(<ErrorComp fail />);
    }).toThrowError('Failed for some reason...');

    expect(spy).not.toHaveBeenCalled();
  });

  describe('root()', () => {
    it('returns an element for a host component', () => {
      const { root } = render(<main />);

      expect(root).toBeInstanceOf(Element);
      expect(root).toBeElementType('main');
    });

    it('returns an element for a class component', () => {
      const { root } = render<TestProps>(<ClassComp />);

      expect(root).toBeInstanceOf(Element);
      expect(root).toBeElementType(ClassComp);
    });

    it('returns an element for a function component', () => {
      const { root } = render<TestProps>(<FuncComp />);

      expect(root).toBeInstanceOf(Element);
      expect(root).toBeElementType(FuncComp);
    });

    it('returns the passed element when using `strict`', () => {
      const { root } = render<TestProps>(<FuncComp />, { strict: true });

      expect(root).toBeInstanceOf(Element);
      expect(root).toBeElementType(FuncComp);
    });

    it('returns the passed element when using `wrapper`', () => {
      const { root } = render<TestProps>(<FuncComp />, { wrapper: <Wrapper /> });

      expect(root).toBeInstanceOf(Element);
      expect(root).toBeElementType(FuncComp);
    });

    it('returns the passed memoized element when using `wrapper`', () => {
      const Root = React.memo(FuncComp);
      const { root } = render<TestProps>(<Root />, { wrapper: <Wrapper /> });

      expect(root).toBeInstanceOf(Element);
      expect(root).toBeElementType(FuncComp);
    });

    it('returns the passed element when using `strict` and `wrapper`', () => {
      const { root } = render<TestProps>(<FuncComp />, { strict: true, wrapper: <Wrapper /> });

      expect(root).toBeInstanceOf(Element);
      expect(root).toBeElementType(FuncComp);
    });
  });

  describe('toJSON()', () => {
    it('returns JSON for a host component', () => {
      const json = render(<main />).toJSON();

      expect(json).toEqual(expect.objectContaining({ type: 'main' }));
    });

    it('returns JSON for a class component', () => {
      const json = render<TestProps>(<ClassComp />).toJSON();

      expect(json).toEqual(expect.objectContaining({ type: 'div' }));
    });

    it('returns JSON for a function component', () => {
      const json = render<TestProps>(<FuncComp />).toJSON();

      expect(json).toEqual(expect.objectContaining({ type: 'span' }));
    });
  });

  describe('toTree()', () => {
    it('returns a tree for a host component', () => {
      const tree = render(<main />).toTree();

      expect(tree).toEqual(expect.objectContaining({ nodeType: 'host', type: 'main' }));
    });

    it('returns a tree for a class component', () => {
      const tree = render<TestProps>(<ClassComp />).toTree();

      expect(tree).toEqual(expect.objectContaining({ nodeType: 'component', type: ClassComp }));
    });

    it('returns a tree for a function component', () => {
      const tree = render<TestProps>(<FuncComp />).toTree();

      expect(tree).toEqual(expect.objectContaining({ nodeType: 'component', type: FuncComp }));
    });
  });

  describe('toString()', () => {
    it('returns name of host component', () => {
      const name = render(<main />).toString();

      expect(name).toBe('<main />');
    });

    it('returns name of class component', () => {
      const name = render<TestProps>(<ClassComp />).toString();

      expect(name).toBe('<ClassComp />');
    });

    it('returns name of function component', () => {
      const name = render<TestProps>(<FuncComp />).toString();

      expect(name).toBe('<FuncComp />');
    });
  });

  describe('mounting', () => {
    describe('class component', () => {
      it('triggers `componentWillMount` and `componentDidMount`', () => {
        const spy = jest.fn();

        class MountTest extends React.Component {
          UNSAFE_componentWillMount() {
            spy();
          }

          componentDidMount() {
            spy();
          }

          render() {
            return null;
          }
        }

        render<{}>(<MountTest />);

        expect(spy).toHaveBeenCalledTimes(2);
      });

      it('supports async `componentDidMount`', async () => {
        const spy = jest.fn();

        const { root } = await renderAndWait<AsyncProps>(<AsyncCdmComp onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(root).toContainNode('Loaded');
      });

      it('supports `componentDidMount` with timers', async () => {
        const spy = jest.fn();

        const { root } = await renderAndWait<AsyncProps>(<TimerCdmComp onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(root).toContainNode('Loaded');
      });
    });

    describe('function component ', () => {
      it('triggers `useEffect` mount on hook', () => {
        const spy = jest.fn();

        function MountTest() {
          useEffect(spy);

          return null;
        }

        render<{}>(<MountTest />);

        expect(spy).toHaveBeenCalledTimes(1);
      });

      it('triggers `useLayoutEffect` mount on hook', () => {
        const spy = jest.fn();

        function MountTest() {
          useLayoutEffect(spy);

          return null;
        }

        render<{}>(<MountTest />);

        expect(spy).toHaveBeenCalledTimes(1);
      });

      it('supports async `useEffect`', async () => {
        const spy = jest.fn();

        const { root } = await renderAndWait<AsyncProps>(<AsyncHookComp onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(root).toContainNode('Loaded');
      });

      it('supports `useEffect` with timers', async () => {
        const spy = jest.fn();

        const { root } = await renderAndWait<AsyncProps>(<TimerHookComp onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(root).toContainNode('Loaded');
      });
    });
  });

  describe('unmounting', () => {
    describe('class component', () => {
      it('triggers `componentWillUnmount`', () => {
        const spy = jest.fn();

        class UnmountTest extends React.Component {
          componentWillUnmount() {
            spy();
          }

          render() {
            return null;
          }
        }

        const result = render<{}>(<UnmountTest />);

        result.unmount();

        expect(spy).toHaveBeenCalledTimes(1);
      });
    });

    describe('function component ', () => {
      it('triggers `useEffect` unmount on hook', () => {
        const spy = jest.fn();

        function UnmountTest() {
          useEffect(() => {
            return spy;
          });

          return null;
        }

        const result = render<{}>(<UnmountTest />);

        result.unmount();

        expect(spy).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('updating', () => {
    let count = 0;

    interface UpdateProps {
      children?: React.ReactNode;
      index?: number;
    }

    class ClassUpdateTest extends React.Component<UpdateProps> {
      render() {
        const { children } = this.props;

        count += 1;

        return children ? <div>{children}</div> : null;
      }
    }

    class PureClassUpdateTest extends React.PureComponent<UpdateProps> {
      render() {
        const { children } = this.props;

        count += 1;

        return children ? <div>{children}</div> : null;
      }
    }

    function FuncUpdateTest({ children }: UpdateProps) {
      count += 1;

      return children ? <div>{children}</div> : null;
    }

    const MemoFuncUpdateTest = React.memo(FuncUpdateTest);

    beforeEach(() => {
      count = 0;
    });

    it('when using `strict`, re-renders the passed element', () => {
      const result = render<TestProps>(<FuncComp name="mount" />, { strict: true });

      expect(result).toMatchSnapshot();
      expect(result.root).toHaveProp('name', 'mount');

      result.update({ name: 'update' });

      expect(result).toMatchSnapshot();
      expect(result.root).toHaveProp('name', 'update');
    });

    it('when using `wrapper`, re-renders the passed element', () => {
      const result = render<TestProps>(<FuncComp name="mount" />, { wrapper: <Wrapper /> });

      expect(result).toMatchSnapshot();
      expect(result.root).toHaveProp('name', 'mount');

      result.update({ name: 'update' });

      expect(result).toMatchSnapshot();
      expect(result.root).toHaveProp('name', 'update');
    });

    it('when using `strict` and `wrapper`, re-renders the passed element', () => {
      const result = render<TestProps>(<FuncComp name="mount" />, {
        strict: true,
        wrapper: <Wrapper />,
      });

      expect(result).toMatchSnapshot();
      expect(result.root).toHaveProp('name', 'mount');

      result.update({ name: 'update' });

      expect(result).toMatchSnapshot();
      expect(result.root).toHaveProp('name', 'update');
    });

    describe('class component', () => {
      it('re-renders if props dont change', () => {
        const result = render<UpdateProps>(<ClassUpdateTest />);

        result.update();
        result.update();

        expect(count).toBe(3);
      });

      it('re-renders if props change', () => {
        const result = render<UpdateProps>(<ClassUpdateTest index={0} />);

        expect(result.root).toHaveProp('index', 0);

        result.update({ index: 1 });

        expect(result.root).toHaveProp('index', 1);

        result.update({ index: 2 });

        expect(result.root).toHaveProp('index', 2);

        expect(count).toBe(3);
      });

      it('re-renders with a different child', () => {
        const result = render<UpdateProps>(<ClassUpdateTest>Foo</ClassUpdateTest>);

        expect(result.root).toContainNode('Foo');

        const child = <div>Bar</div>;

        result.update({}, child);

        expect(result.root).toContainNode(child);
      });

      it('doesnt re-render if pure and props dont change', () => {
        const result = render<UpdateProps>(<PureClassUpdateTest />);

        result.update();
        result.update();
        result.update();
        result.update();
        result.update();

        expect(count).toBe(2);
      });

      it('triggers update life cycles', () => {
        const spy = jest.fn();

        interface Props {
          test: string;
        }

        class UpdateTest extends React.Component<Props> {
          UNSAFE_componentWillReceiveProps(nextProps: Props) {
            expect(nextProps.test).toBe('update');
            spy();
          }

          UNSAFE_componentWillUpdate(nextProps: Props) {
            expect(nextProps.test).toBe('update');
            spy();
          }

          componentDidUpdate() {
            expect(this.props.test).toBe('update');
            spy();
          }

          render() {
            return null;
          }
        }

        const result = render<Props>(<UpdateTest test="mount" />);

        result.update({ test: 'update' });

        expect(spy).toHaveBeenCalledTimes(3);
      });

      it('supports async `componentDidUpdate`', async () => {
        const spy = jest.fn();

        const result = render<AsyncProps>(<AsyncCduComp id="first" onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(0);
        expect(result.root).toContainNode('Loading...');

        await result.updateAndWait({ id: 'second' });

        expect(spy).toHaveBeenCalledTimes(1);
        expect(result.root).toContainNode('Loaded');
      });

      it('supports `componentDidUpdate` with timers', async () => {
        const spy = jest.fn();

        const result = render<AsyncProps>(<TimerCduComp id="first" onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(0);
        expect(result.root).toContainNode('Loading...');

        await result.updateAndWait({ id: 'second' });

        expect(spy).toHaveBeenCalledTimes(1);
        expect(result.root).toContainNode('Loaded');
      });
    });

    describe('function component', () => {
      it('re-renders if props dont change', () => {
        const result = render<UpdateProps>(<FuncUpdateTest />);

        result.update();
        result.update();

        expect(count).toBe(3);
      });

      it('re-renders if props change', () => {
        const result = render<UpdateProps>(<FuncUpdateTest index={0} />);

        expect(result.root).toHaveProp('index', 0);

        result.update({ index: 1 });

        expect(result.root).toHaveProp('index', 1);

        result.update({ index: 2 });

        expect(result.root).toHaveProp('index', 2);

        expect(count).toBe(3);
      });

      it('re-renders with a different child', () => {
        const result = render<UpdateProps>(<FuncUpdateTest>Foo</FuncUpdateTest>);

        expect(result.root).toContainNode('Foo');

        const child = <div>Bar</div>;

        result.update({}, child);

        expect(result.root).toContainNode(child);
      });

      it('doesnt re-render if memoized and props dont change', () => {
        const result = render<UpdateProps>(<MemoFuncUpdateTest />);

        result.update();
        result.update();
        result.update();
        result.update();
        result.update();

        expect(count).toBe(2);
      });

      it('triggers `useEffect` each update when no cache', () => {
        const spy = jest.fn();

        interface Props {
          test: string;
        }

        function UpdateTest({ test }: Props) {
          useEffect(spy);

          return null;
        }

        const result = render<Props>(<UpdateTest test="update" />);

        result.update({ test: 'update' });
        result.update({ test: 'update' });

        expect(spy).toHaveBeenCalledTimes(3);
      });

      it('triggers `useEffect` each update with caching and props change', () => {
        const spy = jest.fn();

        interface Props {
          test: string;
        }

        function UpdateTest({ test }: Props) {
          useEffect(spy, [test]);

          return null;
        }

        const result = render<Props>(<UpdateTest test="mount" />);

        result.update({ test: 'mount' });
        result.update({ test: 'update' });
        result.update({ test: 'update' });

        expect(spy).toHaveBeenCalledTimes(2);
      });

      it('supports async `useEffect`', async () => {
        const spy = jest.fn();

        const result = render<AsyncProps>(<AsyncHookComp id="first" onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(0);
        expect(result.root).toContainNode('Loading...');

        await result.updateAndWait({ id: 'second' });

        expect(spy).toHaveBeenCalledTimes(1);
        expect(result.root).toContainNode('Loaded');
      });

      it('supports `useEffect` with timers', async () => {
        const spy = jest.fn();

        const result = render<AsyncProps>(<AsyncHookComp id="first" onLoad={spy} />);

        expect(spy).toHaveBeenCalledTimes(0);
        expect(result.root).toContainNode('Loading...');

        await result.updateAndWait({ id: 'second' });

        expect(spy).toHaveBeenCalledTimes(1);
        expect(result.root).toContainNode('Loaded');
      });
    });
  });

  describe('re-rendering', () => {
    it('can create multiple root elements (assigned to different vars)', () => {
      const { root, rerender } = render(<div />);
      const spanRoot = rerender(<span />);
      const sectionRoot = rerender(<section />);

      expect(spanRoot).not.toBe(root);
      expect(sectionRoot).not.toBe(root);
      expect(sectionRoot).toBeElementType('section');
    });

    it('can update the wrapper', () => {
      const result = render(<div />, {
        wrapper: <div id="first" />,
      });

      const out1 = result.debug({ log: false });

      expect(out1).toMatchSnapshot();

      result.rerender(<span />, {
        wrapper: <section id="second" />,
      });

      const out2 = result.debug({ log: false });

      expect(out2).toMatchSnapshot();
      expect(out1).not.toBe(out2);
    });

    describe('sync', () => {
      it('can replace the root element', () => {
        const { root, rerender } = render(<div />);

        expect(root).toBeElementType('div');

        const newRoot = rerender(<span />);

        expect(newRoot).toBeElementType('span');
      });

      it('can reuse the root on the result', () => {
        const result = render(<div />);

        expect(result.root).toBeElementType('div');

        result.rerender(<span />);

        expect(result.root).toBeElementType('span');
      });
    });

    describe('async', () => {
      it('can replace the root element', async () => {
        const { root, rerenderAndWait } = await renderAndWait(<div />);

        expect(root).toBeElementType('div');

        const newRoot = await rerenderAndWait(<span />);

        expect(newRoot).toBeElementType('span');
      });

      it('can reuse the root on the result', async () => {
        const result = await renderAndWait(<div />);

        expect(result.root).toBeElementType('div');

        await result.rerenderAndWait(<span />);

        expect(result.root).toBeElementType('span');
      });
    });
  });
});
