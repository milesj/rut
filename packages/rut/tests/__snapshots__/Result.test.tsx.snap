// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result can serialize as a snapshot 1`] = `
<FuncComp>
  <span>
    <b>Child</b>
  </span>
</FuncComp>
`;

exports[`Result re-rendering can update the wrapper 1`] = `
"<div id=\\"first\\">
  <div />
</div>"
`;

exports[`Result re-rendering can update the wrapper 2`] = `
"<section id=\\"second\\">
  <span />
</section>"
`;

exports[`Result updating when using \`strict\` and \`wrapper\`, re-renders the passed element 1`] = `
<Wrapper>
  <FuncComp name="mount">
    <span />
  </FuncComp>
</Wrapper>
`;

exports[`Result updating when using \`strict\` and \`wrapper\`, re-renders the passed element 2`] = `
<Wrapper>
  <FuncComp name="update">
    <span />
  </FuncComp>
</Wrapper>
`;

exports[`Result updating when using \`strict\`, re-renders the passed element 1`] = `
<FuncComp name="mount">
  <span />
</FuncComp>
`;

exports[`Result updating when using \`strict\`, re-renders the passed element 2`] = `
<FuncComp name="update">
  <span />
</FuncComp>
`;

exports[`Result updating when using \`wrapper\`, re-renders the passed element 1`] = `
<Wrapper>
  <FuncComp name="mount">
    <span />
  </FuncComp>
</Wrapper>
`;

exports[`Result updating when using \`wrapper\`, re-renders the passed element 2`] = `
<Wrapper>
  <FuncComp name="update">
    <span />
  </FuncComp>
</Wrapper>
`;

exports[`Result wraps with \`StrictMode\` when using \`strict\` 1`] = `<StrictComp />`;

exports[`Result wraps with another element when using \`wrapper\` 1`] = `
<Wrapper>
  <ContextComp>
    <div>wrapped</div>
  </ContextComp>
</Wrapper>
`;

exports[`Result wraps with both \`strict\` and \`wrapper\` 1`] = `
<Wrapper>
  <Wrapped>
    <div>
      <StrictComp />
    </div>
  </Wrapped>
</Wrapper>
`;
